### DISCUSSION TITLE
boundary condition

### DISCUSSION THREAD

**Discussion by amirhosseinrezaei9776** [Yihao-Shi/GeoTaichi#24] [Q&A]
Dear Yihao-Shi,

I recently came across your impressive work with GeoTaichi, and I wanted to express my admiration for your contributions to the field.

I have been experimenting with your code and created a simple example based on the column collapse scenario, with some parameter modifications. However, I encountered the following error during my simulation:

RuntimeError: Error:: /max_velocity_constraint/ is set as zero!

I would greatly appreciate any guidance or assistance you could provide to help me resolve this issue. Thank you in advance.
here is the fully code:

`from geotaichi import *

init(device_memory_GB=4)

mpm = MPM()

mpm.set_configuration(domain=ti.Vector([20., 20., 20.]),
                      dimension="3-Dimension",
                      background_damping=0.02, 
                      gravity=ti.Vector([0., 0., -9.8]),
                      alphaPIC=0.005, 
                      mapping="USF",
                      stabilize=None,
                      shape_function="GIMP",
                      material_type="Solid")

mpm.set_solver(solver={
                           "Timestep":                   0.1,
                           "SimulationTime":             5.,
                           "SaveInterval":               0.1
                      })

mpm.memory_allocate(memory={
                                "max_material_number":    1,
                                "max_particle_number":    1.0e5
                            })

mpm.add_material(model="MohrCoulomb",
                 material={
                               "MaterialID":           1,
                               "Density":              1700.0,
                               "YoungModulus":    10e5,
                               "Cohesion":        1000.,
                               "Friction":        28.
                 })

mpm.add_element(element={
                             "ElementType":               "R8N3D",
                             "ElementSize":               ti.Vector([0.5, 0.5, 0.5]),
                             "Contact":    {
                                                "ContactDetection":                False
                                           }
                        })

mpm.add_region(region=[{
                            "Name": "region1",
                            "Type": "Rectangle",
                            "BoundingBoxPoint": ti.Vector([1., 1., 1.]),
                            "BoundingBoxSize": ti.Vector([10., 10., 10.]),
                            "zdirection": ti.Vector([0., 0., 1.])
                      }])

mpm.add_body(body={
                       "Template": [{
                                       "RegionName":         "region1",
                                       "nParticlesPerCell":  2,
                                       "BodyID":             0,
                                       "MaterialID":         1,
                                       "ParticleStress": {
                                                              "GravityField":     True,
                                                              "InternalStress":   ti.Vector([-0., -0., -0., 0., 0., 0.]),
                                                              "Traction":         {}
                                                         },
                                       "InitialVelocity":ti.Vector([0., 0., 0.]),
                                       "FixVelocity":    ["Free", "Free", "Free"]
                                                         }]})

mpm.add_boundary_condition(boundary=[
                                        {
                                             "BoundaryType":   "VelocityConstraint",
                                             "Velocity":       [0., 0., 0.],
                                             "StartPoint":     [0., 0., 0.],
                                             "EndPoint":       [20., 20., 1.]
                                        },

                                        {    
                                             "BoundaryType":   "VelocityConstraint",
                                             "Velocity":       [0., 0., 0.],
                                             "StartPoint":     [0., 0., 0.],
                                             "EndPoint":       [1., 20., 20.]
                                        },

                                        {
                                             "BoundaryType":   "VelocityConstraint",
                                             "StartPoint":     [0., 0., 0.],
                                             "EndPoint":       [20., 1., 20.],
                                             "Velocity":       [None, 0., None]
                                        },

                                        {
                                             "BoundaryType":   "VelocityConstraint",
                                             "StartPoint":     [0., 0.5, 0.],
                                             "EndPoint":       [20., 1., 20.],
                                             "Velocity":       [None, 0., None]
                                        }
                                    ])

mpm.select_save_data()

mpm.run()
mpm.postprocessing()`

**Comment by Yihao-Shi** [comment]
You should define max_velocity_constraint in the 'memory' dictionary, 

> mpm.memory_allocate(memory={
"max_material_number": 1,
"max_particle_number": 1.0e5
})

please see [DPmaterial](https://github.com/Yihao-Shi/GeoTaichi/blob/main/example/mpm/ColumnCollapse/DPmaterial.py#L25C65-L25C88) for more details.
